File: java/src/main/java/org/revenj/ServicesPluginLoader.java
package org.revenj;

import org.revenj.extensibility.PluginLoader;

import java.io.*;
import java.net.URL;
import java.nio.charset.Charset;
import java.util.*;

final class ServicesPluginLoader implements PluginLoader {

	private static final String PREFIX = "META-INF/services/";
	private static final Charset UTF8 = Charset.forName("UTF-8");

	private final ClassLoader loader;

	ServicesPluginLoader(ClassLoader loader) {
		this.loader = loader != null ? loader : ClassLoader.getSystemClassLoader();
	}

	public <T> List<Class<T>> find(Class<T> manifest) throws Exception {
		String fullName = PREFIX + manifest.getName();
		//TODO: release class loader to avoid locking up jars on Windows
		Enumeration<URL> configs = loader.getResources(fullName);
		List<Class<T>> plugins = new ArrayList<>();
		while (configs.hasMoreElements()) {
			URL url = configs.nextElement();
			lookupServices(manifest, url, plugins);
		}
		return plugins;
	}

	private <T> void lookupServices(Class<T> manifest, URL u, List<Class<T>> plugins) throws Exception {
		try (InputStream stream = u.openStream();
			 BufferedReader reader = new BufferedReader(new InputStreamReader(stream, UTF8))) {
			String line;
			while ((line = reader.readLine()) != null) {
				int ci = line.indexOf('#');
				if (ci >= 0) line = line.substring(0, ci);
				line = line.trim();
				int n = line.length();
				if (n != 0) {
					if ((line.indexOf(' ') >= 0) || (line.indexOf('\t') >= 0)) {
						throw new IOException("Invalid configuration for " + manifest + " in " + u);
					}
					int cp = line.codePointAt(0);
					if (!Character.isJavaIdentifierStart(cp)) {
						throw new IOException("Invalid configuration for " + manifest + " in " + u);
					}
					for (int i = Character.charCount(cp); i < n; i += Character.charCount(cp)) {
						cp = line.codePointAt(i);
						if (!Character.isJavaIdentifierPart(cp) && (cp != '.')) {
							throw new IOException("Invalid configuration for " + manifest + " in " + u);
						}
					}
					Class<?> service = Class.forName(line);
					plugins.add((Class<T>) service);
				}
			}
		}
	}
}

File: scala/revenj-core/src/main/scala/net/revenj/ServicesPluginLoader.scala
package net.revenj

import java.io.{BufferedReader, IOException, InputStreamReader}
import java.net.{URL, URLEncoder}
import java.nio.charset.Charset

import net.revenj.extensibility.PluginLoader

import scala.collection.mutable.ArrayBuffer
import scala.reflect.runtime.universe._

private[revenj] class ServicesPluginLoader(loader: ClassLoader) extends PluginLoader {
  private val PREFIX: String = "META-INF/services/"
  private val UTF8: Charset = Charset.forName("UTF-8")
  private val mirror = runtimeMirror(loader)

  def find[T: TypeTag]: Seq[Class[T]] = {
    val plugins = new ArrayBuffer[Class[T]]()
    val scalaType = typeOf[T].dealias
    val javaTypeName = Utils.findType(scalaType, mirror).map(_.getTypeName).getOrElse(scalaType.toString)

    val fullName = PREFIX + URLEncoder.encode(javaTypeName, "UTF-8")
    val manifest = mirror.runtimeClass(typeOf[T].dealias).asInstanceOf[Class[T]]
    //TODO: release class loader to avoid locking up jars on Windows
    val configs = loader.getResources(fullName)
    while (configs.hasMoreElements) {
      val url = configs.nextElement
      lookupServices[T](manifest, url, plugins)
    }
    plugins
  }

  private def lookupServices[T](manifest: Class[T], u: URL, plugins: ArrayBuffer[Class[T]]): Unit = {
    val stream = u.openStream
    val reader = new BufferedReader(new InputStreamReader(stream, UTF8))
    try {
      var line: String = null
      while ({line = reader.readLine(); line != null}) {
        val ci = line.indexOf('#')
        if (ci >= 0) line = line.substring(0, ci)
        line = line.trim
        val n = line.length
        if (n != 0) {
          if ((line.indexOf(' ') >= 0) || (line.indexOf('\t') >= 0)) throw new IOException("Invalid configuration for " + manifest + " in " + u)
          var cp = line.codePointAt(0)
          if (!Character.isJavaIdentifierStart(cp)) throw new IOException("Invalid configuration for " + manifest + " in " + u)
          var i = Character.charCount(cp)
          while (i < n) {
            cp = line.codePointAt(i)
            if (!Character.isJavaIdentifierPart(cp) && (cp != '.')) throw new IOException("Invalid configuration for " + manifest + " in " + u)
            i += Character.charCount(cp)
          }
          val service = loader.loadClass(line)
          plugins += service.asInstanceOf[Class[T]]
        }
      }
    } finally {
      stream.close()
      reader.close()
    }
  }
}
